### ISSUE TITLE
Speed up Post Edit Screen with Caching for Flexible Content Previews

### ISSUE THREAD

**Issue by moritzbappert** [acf-extended/ACF-Extended#14]
**Is your feature request related to a problem? Please describe.**
As all of us know, the post edit screen can get quite slow when lots of ACF fields, especially flexible content fields, are used. This is not a problem of your plugin but ACF itself. Nevertheless, I think we might be able to improve this via ACF Extended.

**Describe the solution you'd like**
I've come up with the following general idea, without exactly knowing whether this is possible:

**A)** Instead of waiting for all the flexible content previews to load asynchronously (running PHP and database queries), I would like to instantly see cached previews. 

**How could we achieve this?**
I've thought of these potential ways to solve this:
- save the rendered flexible content previews 
	- with all their HTML, JS and CSS into static HTML files or
	- into local storage / session storage or
	- as images into static files
- initially load all previews from cache
- update the cache when changes are made (what would happen if changes were made elsewhere? How would we invalidate the cache?) 

**B)** Is loading all the meta values for each ACF field one reason for the performance problems?
If so, could we prevent loading all the individual field values initially and only load them on demand, when someone wants to edit one of the flexible content layouts (having only the preview available initially)? 

**C)** Do you have further ideas, how we could improve the performance of the post edit screen?

**Describe alternatives you've considered**

**Additional context**

**Comment by acf-extended** [comment]
Hello @moritzbappert!

Just to let everyone know, the discussion about this topic is ongoing on the [ACF Slack Community](https://join.slack.com/t/wpacf/shared_invite/enQtODc2MjczMzM3NzQ1LTQxNmY2ZGY3OTU2NDkxNTliYmNhMmY1NDMwZGM4NmYxNzgwZTI5MDUzMDFlMGQ5YzcwNDc0ZWM0MDVmODI4NmI).

For history, here is my answer:

Sorry for the late answer, I was kinda busy lately! I've read your suggestion, but haven't had the time to answer yet, because the topic is kinda complex.

I like the idea of cache solution for the preview mode, however it imply a lot of work & tests (as it will store HTML files in the theme or elsewhere, and I have to give as much control as possible to the developer)

It's absolutely possible, however, there's another performance bottlenecks which are more important to me (and should be investigated in my opinion).

There are 3 primary technical limitations which will make Flexible Content outdated compared to Gutenberg in the future:

- All layouts are loaded in the HTML during the admin page load (and hidden in CSS). Those are the models which are duplicated in JS everytime your add a layout. This cause performances issues if you have a very complex Flexible Content (like Page Builder)

- This behavior doesn't allow infinite virtual layouts (a layout that clone itself. For example: Layout row > column > row > column > row etc...)

- Each fields are saved in a separated post_meta (twice due to ACF reference key logic. But this will be fixed in the future, confirmed by Eliott). That means if you have a complicated flexible content with many settings for each layout, they will all be saved in a post meta

For example: If you have 10 common settings for each layout (let's say: margin-top, margin-bottom, margin-left, margin-right, ID, class, padding-top, padding-left, padding-bottom, padding-right). That means you'll have 20 post meta created for each layouts (even if empty). Let's say you have 7 layouts in your page, that makes 140 post meta just for "common layout settings" (without actual content).

Additionally, this will also potentially trigger the PHP maximum post fields limitation, as there could be potentially hundreds of input text/select/radio/number (This can be tweaked in the php.ini to allow PHP to save all $_POST fields during the page save).

Now the question is: Should I bother to tweak & twist the Flexible Content to fix this issues, and basically make it work like Gutenberg. I'm still thinking about it because, in the end, Gutenberg will eat everything. WP core devs already confirmed that everything will be gutenberg in WP (Header, footer, menus, sidebar/widget, pages. Everything).

This is a discussion I would like to have with Eliott, I would like to know what he thinks about it

**Comment by acf-extended** [comment]
Hello,

After some hard work, the initial preview loading will now be processed during page load and not via ajax. As you suggested this increase the page speed load.

This enhancement comes with the full rework & performance boost of the Flexible Content, which will be released in the next patch!

Regards.
